gateway.py

```diff
@@ -156,75 +92,38 @@ async def rx_message_dispatcher(self):
-    async def handle_read_message(self, data):
-        coap_request = aiocoap.Message.decode(data)
-        if coap_request.opt.proxy_uri is None:
-            logger.warning("Proxy-URI option missing in request")
-            response = aiocoap.Message(code=aiocoap.BAD_OPTION)
-        else:
-            target_uri = coap_request.opt.proxy_uri
-            logger.info(f"Proxying request to {target_uri}")
-
-            # Forward the CoAP request to the actual server
-            protocol = await aiocoap.Context.create_client_context()
-            proxy_request = aiocoap.Message(code=coap_request.code, uri=target_uri, payload=coap_request.payload)
-
-            try:
-                coap_response = await protocol.request(proxy_request).response
-                logger.info(f"Forwarded request to {target_uri}, received response: {coap_response} payload: "
-                            f"'{coap_response.payload}'")
-                response = coap_response
-            except Exception as e:
-                logger.error(f"Failed to get response from server: {e}")
-                response = aiocoap.Message(code=aiocoap.INTERNAL_SERVER_ERROR)
-
-        response.mid = coap_request.mid
-        response.token = coap_request.token
-        return response.encode()
-
-    async def handle_write_message(self, data):
-        ...
-
-    def is_valid(self, data) -> bool:
-        try:
-            message = aiocoap.Message.decode(data)
-            # check code type is Unknown if parsing passes
-            return "Unknown" not in message.code.name_printable
-        except:
-            return False
-
+class IoTGatewayServerProtocolTemplate(transport.QUICGatewayServerProtocol):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
 
-class IoTGatewayServerProtocol(transport.QUICGatewayServerProtocol):
     async def tx_message_dispatcher(self):
         logger.info("TX Dispatcher started")
 
     async def rx_message_dispatcher(self):
-        coap_server_context = CoAPServerContext()
         logger.info("RX Dispatcher started")
         while True:
             try:
                 stream_id, data = await self.get_data()
                 logger.info(f"RX Dispatcher - {stream_id}: {data}")
 
-                if coap_server_context.is_valid(data):
-                    response = await coap_server_context.handle_read_message(data)
+                if self.coap_context.is_valid(data):
+                    response = await self.coap_context.handle_read_message(data)
                     await self.send_data(stream_id, response)
+
+                if self.mqtt_sn_context.is_valid(data):
+                    await self.mqtt_sn_context.handle_read_message(data)
             except Exception as e:
                 logger.error("RX Dispatcher - error in received data...")
                 logger.exception(e)
+
+
+def iot_gateway_server_protocol_factory(coap_context, mqtt_sn_context, **kwargs):
```

Comment:
cool stuff, works well for creating duplicate classes with application protocol context as members